### DISCUSSION TITLE
[Q] Logging of `direct` rules

### DISCUSSION THREAD

**Discussion by lonix1** [firewalld/firewalld#1127] [Q&A] [answered]
I want to rate limit ssh connections, so followed the advice in #70.

My config:
````sh
$ sudo firewall-cmd --add-port=2222/tcp

$ sudo firewall-cmd --direct --add-rule ipv4 filter INPUT_direct 0 \
  -p tcp --dport 2222 \
  -m state --state NEW \
  -m recent --name ssh --set

$ sudo firewall-cmd --direct --add-rule ipv4 filter INPUT_direct 1 \
  -p tcp --dport 2222 \
  -m state --state NEW \
  -m recent --name ssh --update --seconds 61 --hitcount 4 --rttl \
  -j REJECT --reject-with tcp-reset

# ...similarly for ipv6
````

That works. If I try ssh 4 times in a minute, I get throttled.

But I also want to log those rejects. So I added `LogDenied=all` to the config.

When I check the logs - `sudo journalctl --since today --identifier kernel` - it shows all the rejected connections, but not the ones from my `direct` rules.

How can I make it log my `direct` rules? Is that supported?

**Comment by erig0** [comment, marked as answer]
> How can I make it log my direct rules? Is that supported?

Not natively. You have to use the iptables log extension, e.g. `-j LOG`. firewalld's `--direct` is mostly a passthrough to iptables.

Unfortunately you'll need two direct rules as iptables doesn't support `-j LOG -j ACCEPT`.

**Comment by lonix1** [comment]
Thanks to @erig0 who put me in the right direction.... for anyone else who needs this, one must add another rule, identical to the REJECT or DROP rule, but which jumps to the "non-terminating" [LOG target](https://linux.die.net/man/8/iptables).

````sh
$ sudo firewall-cmd --direct --add-rule ipv4 filter INPUT_direct 0 \
  -p tcp --dport 22 \
  -m state --state NEW \
  -m recent --name ssh --set \
  -m comment --comment "limit ssh connections per ip"

$ sudo firewall-cmd --direct --add-rule ipv4 filter INPUT_direct 1 \
  -p tcp --dport 22 \
  -m state --state NEW \
  -m recent --name ssh --update --seconds 61 --hitcount 4 --rttl \
  -m limit --limit "5/min" \
  -j LOG --log-prefix "[REJECT SSH BRUTE FORCE] " --log-level 6 \    # <--------------
  -m comment --comment "limit ssh connections per ip"

$ sudo firewall-cmd --direct --add-rule ipv4 filter INPUT_direct 2 \
  -p tcp --dport 22 \
  -m state --state NEW \
  -m recent --name ssh --update --seconds 61 --hitcount 4 --rttl \
  -j REJECT --reject-with tcp-reset \
  -m comment --comment "limit ssh connections per ip"

# ...similarly for ipv6
````

To view logs for those connections:
````sh
# all rejected and dropped
$ sudo journalctl --grep 'kernel'

# or just just ssh brute force attempts
$ sudo journalctl --grep 'REJECT SSH BRUTE FORCE'
````
